MemberRepository 테스트 작성

vectorch9 opened the pull request:
<!--
✅ Resolve: #이슈번호 형태로 입력해 주세요.
ex) Resolve: #123
-->
## 📌Linked Issues
- Resolve: #190 

<!--
✅ 변경 사항을 자세히 알려주세요. (why -> what -> how)
-->
## ✏Change Details
### 테스트 관련 유틸 추가
- PostUtils 추가
- Post와 연관된 Utils도 추가
  - PostTag, PostImage, Comment
### 테스트 작성
- MemberRepository내 `메서드 이름을 통한 쿼리`, 두 개의 Querydsl에 대한 테스트를 각각 작성
### 중복되는 역할의 메서드 삭제
- MemberRepository내 `findById`, `findByUsernames` 두개의 메서드 삭제
  - 전자의 경우 JpaRepository에서 이미 지원하고, 후자의 경우 `findAllByUsernameIn`과 동일한 로직이라 삭제했습니다.
- 추가로, `MemberPostRepositoryQuerydslImpl`에 inner join을 적용 가능한 경우, 적용 하였습니다.


<!--
✅ 추가로 전달할 내용이 있다면 적어주세요.
-->
## 💬Comment
<!--
✅ 참고한 사이트가 있다면 공유해주세요.
-->
- 본격적으로 작성하는 첫 테스트라 작성하면서도 긴가민가한 부분이 많네요.. 피드백 주시면 수정하겠습니다
- 테스트 내 prepare 메서드가 있는데 `@beforeAll` 같은 어노테이션으로 빼고 싶었는데 member를 파라미터로 받기 때문에 어렵네요. 뭔가 좋은 방법이 있을까요? 아니면 그대로 두는게 맞을까요?
- 그 외에도 중복되는 작업이 좀 있는데 이부분 따로 빼지 않는게 나을 것 같다고 생각했습니다. (assert 부분이나 postCount와 같은 변수 선언 부분)
## 📑References
- 

## ✅Check List
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?

vectorch9 committed:
Fix: 의미 없는 findById 제거

- findById는 JPA가 제공하기 때문에 불필요

vectorch9 committed:
Test: MemberRepository 테스트 작성
- 메서드 명에 의해 생성된 쿼리 테스트

vectorch9 committed:
Refactor: MemberRepository Querydsl 관련 리팩토링
- 중복되는 로직을 갖는 메서드 삭제
- 중복되는 빈줄 삭제

vectorch9 committed:
Test: Post 인스턴스 생성 유틸 클래스 추가

vectorch9 committed:
Test: MemberRepositoryQuerydsl 테스트 코드 작성

vectorch9 committed:
Test: Post와 연관된 엔티티 관련 Utils 작성

vectorch9 committed:
Test: MemberPostRepositoryQuerydsl 테스트 작성

vectorch9 committed:
Refactor: MemberPostRepositoryQuerydslImpl에 inner join 적용

vectorch9 committed:
Test: 의미 없는 파라매터 제거

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,119 @@
+	@Autowired
+	private MemberRepository memberRepository;
+
+	@Test
+	void findByUsername_MemberExist_ReturnMember() {
+		// given
+		final Member givenMember = MemberUtils.newInstance();
+		memberRepository.save(givenMember);
+
+		// when
+		final Optional<Member> member = memberRepository.findByUsername(givenMember.getUsername());
+
+		// then
+		assertThat(member.isPresent()).isTrue();
+	}
```
Optional이 본래 목적과 다르게 사용되는 것 같습니다. Optional을 필드로 사용하기 보다는, 체이닝 메소드를 이용해서 boolean 값을 사용하는 것을 제안드립니다.
```java
final boolean isPresent = memberRepository.findByUsername(givenMember.getUsername()).isPresent();
```
> 저희 비즈니스 코드에도 Optional을 잘못 사용하는 부분이 많은데, 리팩토링이 필요해 보이네요.

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,119 @@
+		assertThat(memberExists).isFalse();
+	}
+
+	@Test
+	void findAllByUsernameIn_3MembersExist_Find3Members() {
+		// given
+		final Set<String> usernames = new HashSet<>();
+		while (usernames.size() < 3) {
+			final Member givenMember = MemberUtils.newInstance();
+			if (usernames.contains(givenMember.getUsername())) {
+				continue;
+			}
+			memberRepository.save(givenMember);
+			usernames.add(givenMember.getUsername());
+		}
+
+		// when
+		final List<Member> members = memberRepository.findAllByUsernameIn(usernames);
+
+		// then
+		assertThat(members.size()).isEqualTo(usernames.size());
+	}
```
의미가 불분명한 매직 넘버를 상수로 선언하면 좀 더 가독성있는 코드가 될 것 같습니다.

vectorch9 replied:
다른 테스트에선 따로 상수로 선언했는데 이 파일만 그러네요.. 수정하겠습니다

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,119 @@
+		// given
+		final Set<String> usernames = new HashSet<>();
+		final Set<Long> ids = new HashSet<>();
+		while (usernames.size() < 3) {
+			final Member givenMember = MemberUtils.newInstance();
+			if (usernames.contains(givenMember.getUsername())) {
+				continue;
+			}
+			memberRepository.save(givenMember);
+			usernames.add(givenMember.getUsername());
+			ids.add(givenMember.getId());
+		}
```
이 부분은 공통 메소드로 뽑아낼 수 있을 것 같아요.
MemberUtils에서 username이 서로 다른 List<Member>를 반환하는 static 메소드를 만들어 사용하면 어떨까요?

vectorch9 replied:
넵 이 부분은 추가하겠습니다

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,119 @@
+	void findAllByIdIn_3MembersExist_Find3Members() {
+		// given
+		final Set<String> usernames = new HashSet<>();
+		final Set<Long> ids = new HashSet<>();
```
ids를 꼭 Set 자료구조에 저장해야 할까요?

vectorch9 replied:
음.. List로 바꾸겠습니다.

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberPostRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,170 @@
+
+		// then
+		assertThat(memberPostDtoPage.getContent().size()).isEqualTo(postCount);
+		assertThat(memberPostDtoPage.getContent().get(0).isHasManyPosts()).isTrue();
+		assertThat(memberPostDtoPage.getContent().get(0).getPostCommentsCount()).isEqualTo(postCommentCount);
+		assertThat(memberPostDtoPage.getContent().get(0).getPostLikesCount()).isEqualTo(postLikeCount);
+		assertThat(memberPostDtoPage.getContent().get(0).isPostLikeFlag()).isFalse();
```
메소드 이름에서 기대하는 결과보다 너무 많은 것들을 테스트하고 있는 것 같아요.

회원이 3개의 게시물을 업로드한 상태에서, 3개의 게시물을 조회하는 결과를 확인하는 테스트에서 이러한 것들도 검증하는게 꼭 필요할까요? 필요하다고 생각된다면, 따로 작은 단위로 테스트하는 것이 좋을 것 같아요.

vectorch9 replied:
예를들어 쿼리에서 postCommentCount가 정상적으로 받아져왔는지 테스트 하려면 어떤 식으로 분리를 해야할까요? 아니면 테스트가 필요가 없는 부분일까요 ?
추가로, 분리를 해야 한다면 memberPost의 commentCount, memberTaggedPost의 commentCount, memberSavedPost의 commentCount를 각각 테스트해야 할까요?

seonpilKim replied:
테스트 메소드 네이밍을 기준으로 봤을 때, 어느 범위까지 테스트하는게 한 눈에 파악이 가능할지를 고려하면 좋을 것 같아요.

게시물 댓글 수 검증이 필요 없다는 뜻이 아니라, 3개의 회원 게시물을 조회하는 결과를 기대하는 메소드에 다른 결과들까지 테스트하고 있는 게 과하지 않나 하는 뜻이었어요. 메소드명을 적절하게 지어서 몇몇 부분을 묶어서 테스트하거나, 작은 단위로 쪼개서 테스트하는 편이 가독성 측면에서도 좋을 것 같습니다.

업로드 게시물 / 태그된 게시물 / 저장한 게시물 모두 각각 테스트하는게 좋다고 생각하고, 비슷한 로직은 공통 메소드로 잘 묶을 수 있을 것 같네요.

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberPostRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,170 @@
+		preparePosts(member, postCount, postImageCount, postCommentCount, postLikeCount);
+
+		// when
+		Pageable pageable = PageRequest.of(0, 15);
+		Page<MemberPostDto> memberPostDtoPage = memberRepository.findMemberPostDtos(member.getId(),
+			member.getUsername(), pageable);
```
final 키워드도 붙여주면 좋을 것 같아요.

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberPostRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,170 @@
+		assertThat(memberPostDtoPage.getContent().get(0).isPostLikeFlag()).isFalse();
+	}
+
+	private void preparePosts(Member member, long postCount, long postImageCount, long postCommentCount,
+		long postLikeCount) {
+		for (long count = 1; count <= postCount; count++) {
+			final Post post = PostUtils.newInstance(member);
+			postRepository.save(post);
+
+			final Bookmark bookmark = Bookmark.builder()
+				.member(member)
+				.post(post)
+				.build();
+			bookmarkRepository.save(bookmark);
+
+			for (long imageCount = 1; imageCount <= postImageCount; imageCount++) {
+				final PostImage postImage = PostImageUtils.newInstance(post);
+				postImageRepository.save(postImage);
+
+				final PostTag postTag = PostTagUtils.newInstance(postImage, member.getUsername());
+				postTagRepository.save(postTag);
+			}
+
+			for (long commentCount = 1; commentCount <= postCommentCount; commentCount++) {
+				final Comment comment = CommentUtils.newInstance(post, member);
+				commentRepository.save(comment);
+			}
+
+			for (long likeCount = 1; likeCount <= postLikeCount; likeCount++) {
+				final Member likeMember = MemberUtils.newInstance();
+				memberRepository.save(likeMember);
+				final PostLike postLike = PostLike.builder()
+					.post(post)
+					.member(likeMember)
+					.build();
+				postLikeRepository.save(postLike);
+			}
+		}
+	}
```
회원이 3개의 게시물을 올린 상태에서, 3개의 게시물이 조회되는지 검증하는 메소드에 굳이 image를 여러 개 추가할 필요가 있을까요?
comment, like 이런 것들도 추가하는 건 과하진 않을까요?

공통으로 사용되는 메소드로 사용하기엔 각 단위 테스트마다 불필요한 로직이 많이 끼어있는 것 같아요. bookmark 저장, postTag 저장 등

vectorch9 replied:
생각해보니 반환 값을 List<Post>로 주면 bookmark, tag를 따로 처리할 수 있겠네요.

seonpilKim reviewed (changes requested):
내용이 중복되는 리뷰들은 추가로 comment를 남기진 않았으니 참고바랍니다!

vectorch9 committed:
Test: MemberUtils에 여러 Member를 반환하는 메서드 추가
- 멤버 List에서 username, id를 추출해주는 함수 추가

vectorch9 committed:
Test: MemberUtils에 여러 Member를 반환하는 메서드 추가
- 멤버 List에서 usernames를 추출해주는 함수 추가

vectorch9 committed:
Merge remote-tracking branch 'origin/Test/190' into Test/190

# Conflicts:
#	src/test/java/cloneproject/Instagram/util/domain/member/MemberUtils.java

vectorch9 committed:
Test: MemberRepositoryTest 수정
- Optional는 메서드 체인 사용하도록 수정
- 매직넘버 상수로 전환
- 중복되는 코드를 MemberUtils를 사용하도록 수정

vectorch9 committed:
Test: MemberRepositoryQuerydslTest 메서드 명 수정
- 메서드 명에 WithDetail 추가
- final 키워드 적용 가능한 부분 적용

vectorch9 committed:
Style: MemberPostDto내 인스턴스 변수 명 변경
- hasManyPosts -> hasManyPostImages
  - 실제 변수의 역할에 맞게 이름 변경

vectorch9 committed:
Refactor: 사용하지 않는 클래스 삭제

vectorch9 committed:
Test: Post 연관 엔티티 세부 설정과 관련된 Utils 추가

vectorch9 committed:
Test: MemberPostRepositoryQuerydslTest 역할 분리 및 메서드 추출
- 메서드의 역할 분리
  - 여러개의 포스트가 받아와지는지 / 받아온 포스트의 세부 설정이 잘 반영되어 있는지 두 개의 테스트로 분리
- 앞선 커밋의 Utils로 추출한 메서드 적용

vectorch9 committed:
Test: FollowUtils 추가

- 다수의 Following, Follower관계를 만드는 메서드 추가

vectorch9 committed:
Test: MemberRepositoryQuerydslTest내 메서드 분리 및 추출
- 팔로우관련 코드를 Utils로 추출
- 팔로우 플래그 관련된 테스트를 분리

vectorch9 committed:
Style: 파일 끝 빈줄 추가

vectorch9 commented:
# 추가 변경 사항
## 기존 코드 리팩토링
- 사용하지 않는 PostUtil 삭제
  - 테스트용 Utils가 아닌 기존 비지니스 코드의 Util을 삭제했습니다.
  - Post를 MemberPostDto로 변환해주는 메서드만 갖고있었는데, MemberPostDto 엔티티의 정보뿐 아닌 추가적인 정보를 필요로 하기 때문에 해당 메서드가 사용되지 않고 있었습니다. ( 과거에 MemberPost 관련 리팩토링을 하며 삭제했어야 하는데 안한 것 같아요.)
- MemberPostDto 내 변수 명 수정
  - `hasManyPosts`를 `hasManyPostImages`로 변경
    - 원래의 역할에 맞게 변수 명을 수정하였습니다. 
## 다수의 Utils 추가
- 다른 테스트에서도 사용할 만한 코드를 추출하여 Utils로 작성
- 메서드 명은 최대한 직관적으로 지었습니다
## 테스트 수정사항
### 공통
- Utils로 추출된 메서드를 적용
- 메서드 명을 테스트 하고자 하는 목적을 좀 더 확실히 명시하도록 수정
- 한 메서드에서 너무 많은 부분을 테스트 하고 있는 경우 분리
### MemberPostQuerydsl
- 기존의 prepare 메서드 분리
  - Post의 세부사항(댓글, 이미지, 좋아요)를 준비해주는 메서드
  - bookmark를 준비해주는 메서드
  - tag를 준비해주는 메서드 
- 다수의 포스트 관련 테스트 / 포스트의 세부 사항 관련 테스트로 역할 분리
### MemberQuerydsl
- prepare 관련 메서드 수정
- Follow 플래그 관련 테스트 분리
### Member
- 매직넘버 상수화
- Optional과 관련된 메서드 체인 적용

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberPostRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,248 @@
+		postRepository.saveAll(posts);
+
+		// when
+		final Pageable pageable = PageRequest.of(0, 15);
```
pageable은 given으로 옮기는게 좋을 것 같아요.
0과 15도 각각 상수로 선언하면 어떨까요? `ex) page, size`

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberPostRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,248 @@
+		assertThat(memberPostDtoPage.getContent().get(0).isPostLikeFlag()).isFalse();
+	}
+
+	private List<PostImage> preparePostDetail(Post post, Member member, long postImageCount, long postCommentCount,
```
메소드명만 봐서는 어떤 기능인지 이해하기 힘든 것 같아요.
네이밍을 조금 더 자세하게 변경해보면 어떨까요?
`ex) setPostImageAndCommentAndLike`

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberPostRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	@Test
+	void findMemberPostDtos_PostHasDetail_FindMemberPostDtoWithDetail() {
```
select 절의 scalar subquery의 결과를 Detail이라 표현하신 걸까요?
Detail이라 하니 정확히 무슨 내용인지 파악하기가 어렵고, 명확한 기준도 잡히지 않은 것 같아서 조금 모호한 것 같아요.

테스트 코드 작성 관련해서 몇 가지 제안 사항을 정리해보자면,

1. Dto 조회의 경우 모든 필드가 원하는 값으로 잘 매핑되었는지 확인하는 테스트 작성
    - 네이밍: `(메소드명)_ValidArguments_AllFieldsMappedSuccess`
    - 사실 각 필드마다 메소드를 분리하는게 좋을 것 같긴 한데, 정적인 데이터들은 매핑 테스트만 하면 충분한 것 같아서, 차라리 모든 필드가 정상적으로 매핑되는 테스트 하나와, 동적 쿼리나 다른 조건이 있는 부분들은 따로 메소드로 뽑아서 테스트하면 좋을 것 같습니다.
2. 동적 쿼리(인수에 따라 결과가 다른 쿼리)는 시나리오별로 따로 분리해서 테스트 작성
    - 예를 들면 select절의 `existPostLikeWherePostEqAndMemberIdEq(Long memberId)`, `post.postImages.size().gt(1)`, where 절의 `post.member.username.eq(username)`, orderBy를 적용한 경우 등
3. 페이징 조회의 경우 추가로 `총 페이지 수, 총 원소 수, 현재 페이지 숫자, 현재 페이지 원소 개수` 테스트 작성
    - 네이밍: `(메소드명)_30(엔티티명)SavedAndPage1AndSize5_Total30(엔티티명)And6PagesAndCurrent5(엔티티명)And1Page`

다른 아이디어도 있으시다면 말씀해주세요 :)

vectorch9 replied:
1, 2 이부분은 적용 하겠습니다.
2. `post.postImages.size().gt(1)`의 경우, `메서드명_PostHaveManyPosts_HasManyPostImagesMappedSuccess`식으로 메서드명을 지으면 될까요?
3. 페이지 부분도 테스트 하고 싶었는데 어떤 방식으로 할지 고민하고 있었는데 말씀주신대로 하면 될 것 같습니다.

vectorch9 replied:
3번의 경우 네이밍을 `(메소드명)_30(엔티티명)SavedAndPage1AndSize5_FindTotal30(엔티티명)And6PagesAndCurrent5(엔티티명)And1Page`식으로 3번째 절 앞에 Find를 붙여주는게 좋을 것 같습니다!

vectorch9 replied:
실제로 값을 넣었을때, `findMemberPostDtos_30PostsSavedAndPage1AndSize15_FindTotal30PostsAnd2PagesAndCurrent15PostsAnd0Page`에서 Page가 두번 들어가기 때문에 한눈에 들어오지 않을 것 같습니다. (앞에 Total과 Current가 있으나 And절로 연결되어있어서 
 알아보기 불편한 것 같아요)
스프링에서는 따라서 각 Page를 표현할 때는 Slice라는 명칭을 사용하는 것 같습니다.

> 현재 페이지 번호를 리턴하는 `page.getNumber()`의 설명 주석을 보시면 `Slice`라는 표현을 사용하고 있습니다. 좀 더 자세히 살펴보면 Page는 Slice를 상속하고 있으나, 상속하며 추가되는 메서드가 주로 Total과 관련된 메서드입니다.

따라서 저희도 Total을 표현할땐 Page를, 각각의 페이지를 표현할 땐 Slice를 사용하면 좋을것 같아요. 
 즉, `findMemberPostDtos_30PostsSavedAndPage1AndSize15_FindTotal30PostsAnd2PagesAndCurrent15PostsAndFirstSlice`의 형태로 네이밍하면 좋을 것 같습니다.

여기서 고민되는 점이 `FirstSlice` 형식으로 작성하는게 의미 상 맞는데, Slice 인덱스가 0부터 시작합니다. 아래의 두 방법 중 어느 방법이 좋을까요?
1. 인덱스는 0부터 세지만 **첫번째** 슬라이스는 맞기 때문에 `FirstSlice`라고 표현 -> 인덱스 1은 `SecondSlice` -> ...
2. 인덱스에 맞게 이름 작성. 인덱스 0 = `ZeroSlice` -> 인덱스1 = `FirstSlice` -> ..

추가로 현재 Slice 넘버를 검증하기 위해
`assertThat(memberPostDtoPage.getNumber()).isEqualTo(1);`와 같은 코드를 짤텐데 여기의 1부분도 given에서 상수로 선언하는게 맞겠죠?

vectorch9 replied:
+ 추가로 조사해보니 쿼리를 Slice로도 날릴 수 있네요.. 네이밍을 어떻게 하는게 좋을까요

seonpilKim replied:
> post.postImages.size().gt(1)의 경우, 메서드명_PostHaveManyPosts_HasManyPostImagesMappedSuccess식으로 메서드명을 지으면 될까요?

Many 보다는 구체적으로 몇 개의 이미지를 가지고 있는지를 명시하는게 좋을 것 같아요.
`_PostHas3Images_HasManyPostImages`


> 3번의 경우 네이밍을 (메소드명)_30(엔티티명)SavedAndPage1AndSize5_FindTotal30(엔티티명)And6PagesAndCurrent5(엔티티명)And1Page식으로 3번째 절 앞에 Find를 붙여주는게 좋을 것 같습니다!

좋아요~


> 실제로 값을 넣었을때, findMemberPostDtos_30PostsSavedAndPage1AndSize15_FindTotal30PostsAnd2PagesAndCurrent15PostsAnd0Page에서 Page가 두번 들어가기 때문에 한눈에 들어오지 않을 것 같습니다. (앞에 Total과 Current가 있으나 And절로 연결되어있어서
알아보기 불편한 것 같아요)
스프링에서는 따라서 각 Page를 표현할 때는 Slice라는 명칭을 사용하는 것 같습니다.
현재 페이지 번호를 리턴하는 page.getNumber()의 설명 주석을 보시면 Slice라는 표현을 사용하고 있습니다. 좀 더 자세히 살펴보면 Page는 Slice를 상속하고 있으나, 상속하며 추가되는 메서드가 주로 Total과 관련된 메서드입니다.
따라서 저희도 Total을 표현할땐 Page를, 각각의 페이지를 표현할 땐 Slice를 사용하면 좋을것 같아요.
즉, findMemberPostDtos_30PostsSavedAndPage1AndSize15_FindTotal30PostsAnd2PagesAndCurrent15PostsAndFirstSlice의 형태로 네이밍하면 좋을 것 같습니다.

저는 Slice라고 표현하니까 오히려 어색한 것 같아요🤔
현재 몇 페이지인지 검증하는 메소드를 굳이 슬라이스라고 바꿔서 표현할 필요가 있을까요?
Page에서 제공하는 getTotalPages() 메소드도 있는 만큼, 현재 페이지 수를 ~Page 라고 표현하는 편이 좋을 것 같아요.
저희가 일상적으로 사용하는 단어도 페이지이기도 하고, 저희 입장에서 자연스러운 표현을 사용하는게 좋다고 생각합니다.

> 여기서 고민되는 점이 FirstSlice 형식으로 작성하는게 의미 상 맞는데, Slice 인덱스가 0부터 시작합니다. 아래의 두 방법 중 어느 방법이 좋을까요?
인덱스는 0부터 세지만 첫번째 슬라이스는 맞기 때문에 FirstSlice라고 표현 -> 인덱스 1은 SecondSlice -> ...
인덱스에 맞게 이름 작성. 인덱스 0 = ZeroSlice -> 인덱스1 = FirstSlice -> ..

저는 숫자로 표기하는게 좋을 것 같습니다.

> 추가로 현재 Slice 넘버를 검증하기 위해
assertThat(memberPostDtoPage.getNumber()).isEqualTo(1);와 같은 코드를 짤텐데 여기의 1부분도 given에서 상수로 선언하는게 맞겠죠?

넵 상수로 선언하지 않으면 어떤 의미인지 파악하기 어려울 것 같아요.

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,222 @@
+		prepareFollow(member, followerCount, followingCount);
+
+		// when
+		final UserProfileResponse userProfileResponse = memberRepository.findUserProfile(-1L, member.getUsername());
```
-1L이 무슨 의미인지 잘 모르겠네요😅 따로 변수로 추출하면 좋을 것 같아요.

seonpilKim reviewed (changes requested):
리뷰가 예상보다 늦어졌네요😂
Detail이라는 네이밍을 여러 곳에서 사용하셨는데, 사실 메소드명만 봐서는 어떤 내용을 테스트하는 건지 감이 잘 안오는 것 같습니다 ㅠㅠ;
가능하다면 조금 더 직관적인 메소드명으로 변경하거나, 단위를 더 작게 나누어서 테스트하는게 좋을 것 같아요.
테스트 코드 작성 관련 제안 사항도 아래 리뷰에 추가로 작성하였으니, 다훈님 의견도 말씀해주시면 감사하겠습니다~!

vectorch9 committed:
Style: Pageable, memberId 관련 매직넘버 상수화

vectorch9 committed:
Style: 메서드명을 자세한 정보를 포함하도록 수정

vectorch9 committed:
Test: MemberPost API 테스트 시 Page 관련 테스트 추가

vectorch9 committed:
Test: MemberPost API 테스트에서 page관련 분리

- 저장, 태그된 게시물 조회 테스트 분리
  - 여러개의 게시물 중 일부만 저장, 태그된 경우에 대한 테스트를 분리해냄

vectorch9 committed:
Test: MemberPost 관련 테스트 세분화
- 멤버가 게시물에 좋아요를 누른 경우에 대한 테스트 추가
- 포스트내 이미지가 여러개 있는 경우에 대한 테스트 추가
- 일부 메서드명 수정

vectorch9 committed:
Test: MemberRepositoryQuerydslTest 세분화
- 팔로우, 차단관련 플래그 테스트 작성
- 매직넘버 상수화

vectorch9 committed:
Style: 첫번째 index(0) 상수화

vectorch9 commented:
## 추가 변경 사항
- 매직넘버 상수화
### 테스트 세분화
- MemberRepositoryQuerydsl이 반환하는 여러 플래그(팔로우, 차단관련)에 대한 테스트 작성
- MemberPostRepositoryQuerydsl 테스트 세분화/분리
  - Page에 관련된 테스트
  - 전체 게시물 중 일부만 업로드/저장/태그된 경우에 대한 테스트
  - 정상적인 요청의 경우 값이 잘 매핑됐는지 테스트
  - hasManyPostImages 플래그에 대한 테스트
  - likeFlag 플래그(해당 게시물에 좋아요를 눌렀는지)에 대한 테스트

vectorch9 committed:
Test: MemberRepositoryQuerydslTest에 default flag가 false인지 검증하는 로직 추가

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,104 @@
+		memberRepository.save(givenMember);
+
+		// when
+		final boolean memberExists = memberRepository.existsByUsername(givenMember.getUsername());
```
boolean 타입 변수명은 `is~~`|`has~~` 으로 사용하면 어떨까요?

vectorch9 replied:
1-1. 이 부분은 저도 마음에 들지 않았는데 좋은 아이디어네요. 단, `MemberBlockedOne`의 경우에 Member가 One을 조회한건 지 One이 Member를 조회한 건지 명시하기 위해 `OneRequestMemberProfile`의 형태가 추가되어야 할텐데 문법상 올바르지 않을 것 같습니다. 차라리`MemberBlockedMeAndIRequestMemberProfile`의 형식으로 만든다면 어떨까요? 아니면 Tom과 Jerry식으로 이름을 지어주는 것도 나쁘지 않을 것 같습니다.
1-2. `1-1` 에서 예를들기위해 Request라는 단어를 사용했는데 사실 이 표현도 별로 마음에 들지 않네요. `A가 B의 프로필을 조회한다.` 라는 조건을 표현하기 위한 좋은 단어가 있을까요? _역시 이름짓는게 제일 힘드네요...._ 
1-3. 현재 메서드 명이 너무 길어져서 가독성이 안좋아지는 것 같은데 좋은 방법이 없을까요? (현재 메서드명이 긴 경우 한 라인에 130~140자 가량 되는 것 같습니다) 컨벤션 상 한줄에 최대 120자 정도로 제한이 되어있을텐데 메서드 명이 긴거라 IDE가 처리해줄 방법이 없는 것 같습니다.
2. 이 부분은 사실 오늘 회의때 말씀드릴려 했는데... 모든 시나리오를 테스트하고 싶은데 현재 파일 가독성이 너무 떨어진다고 생각됩니다. 현재 MemberPost 테스트 코드가 한 파일에 650줄(추가적인 시나리오까지 작성하면 더 길어지겠죠) 가량에 달하는데 각 메서드에 테스트를 정상적으로 작성했는지 여부를 판단하기 힘들 정도로 길어진 것 같습니다. 특히 MemberPost의 경우 똑같은 구조로 3개의 각 메서드를 반복해서 테스트 하다보니 더 심한 것 같아요. 중간에 테스트하는 메서드를 기준으로 구분선을 넣던가 Nested 클래스를 사용하거나 파일을 분리하는 등의 방법은 어떨까요?
Nested 사용은 아래와 같은 형태를 의미합니다
 ```
@Nested
    class update {
        @Test
        void StateA_ResultA() {
                 // 테스트 내용
        }

        @Test
        void StateB_ResultB() {
                 // 테스트 내용
        }
    }
```
> referenced by https://mkyong.com/junit5/junit-5-nested-test-examples/

예를들어 update 메서드를 테스트 하는 경우, 상위 클래스 이름을 테스트할 메서드의 이름과 동일하게 적고 하위 메서드에 여러 시나리오별로  작성하는 형태입니다. 저희의 경우 하위 메서드 이름은 `현재 조건 및 상태_예상되는 결과`로 지으면 될 것 같습니다. 이런 형식을 사용하면 위에서 언급된 메서드명이 매우 길어지는 문제도 어느정도 해결되지 않을까 싶습니다.
혹시 현업에선 테스트코드가 매우 길어지는 경우 어떻게 가독성을 높일까요?

> 검색을 해보아도 관련된 문제에 관한 글을 찾기 힘드네요..

seonpilKim replied:
> 이 부분은 저도 마음에 들지 않았는데 좋은 아이디어네요. 단, MemberBlockedOne의 경우에 Member가 One을 조회한건 지 One이 Member를 조회한 건지 명시하기 위해 OneRequestMemberProfile의 형태가 추가되어야 할텐데 문법상 올바르지 않을 것 같습니다. 차라리MemberBlockedMeAndIRequestMemberProfile의 형식으로 만든다면 어떨까요? 아니면 Tom과 Jerry식으로 이름을 지어주는 것도 나쁘지 않을 것 같습니다.

**(수정)**
필요하다면 메소드 호출 주체와 객체도 네이밍에 포함하면 더 가독성이 좋을 것 같네요.

> 1-1 에서 예를들기위해 Request라는 단어를 사용했는데 사실 이 표현도 별로 마음에 들지 않네요. A가 B의 프로필을 조회한다. 라는 조건을 표현하기 위한 좋은 단어가 있을까요? 역시 이름짓는게 제일 힘드네요....

Request 보다는 그냥 메소드 네이밍에서 일부를 차용하면 어떨까요? 
음 그리고 저는 메소드를 호출하는 입장에서 네이밍을 짓는게 읽기 좋은 것 같은데.. 다훈님은 어떠실지 모르겠네요..
ex) `~~_TomWasBlockedByJerryAndTomFindJerryProfile_~~`

> 현재 메서드 명이 너무 길어져서 가독성이 안좋아지는 것 같은데 좋은 방법이 없을까요? (현재 메서드명이 긴 경우 한 라인에 130~140자 가량 되는 것 같습니다) 컨벤션 상 한줄에 최대 120자 정도로 제한이 되어있을텐데 메서드 명이 긴거라 IDE가 처리해줄 방법이 없는 것 같습니다.

네이밍이 길어져서 120자를 넘는 것은 크게 문제되지 않는다고 생각합니다. 가능하면 짧게 짓는게 좋겠지만, 과도하게 축약하면 가독성이 깨지는 문제가 있으니, 길어지더라도 명확하게 네이밍을 하는게 더 중요하다고 생각해요.

> 이 부분은 사실 오늘 회의때 말씀드릴려 했는데... 모든 시나리오를 테스트하고 싶은데 현재 파일 가독성이 너무 떨어진다고 생각됩니다. 현재 MemberPost 테스트 코드가 한 파일에 650줄(추가적인 시나리오까지 작성하면 더 길어지겠죠) 가량에 달하는데 각 메서드에 테스트를 정상적으로 작성했는지 여부를 판단하기 힘들 정도로 길어진 것 같습니다. 특히 MemberPost의 경우 똑같은 구조로 3개의 각 메서드를 반복해서 테스트 하다보니 더 심한 것 같아요. 중간에 테스트하는 메서드를 기준으로 구분선을 넣던가 Nested 클래스를 사용하거나 파일을 분리하는 등의 방법은 어떨까요?

Nested 클래스 괜찮은 것 같네요. 하나의 메소드에 여러 테스트 코드를 작성하면 메소드 이름을 여러 번 중복 작성하는 번거로움도 줄고, 어디까지가 해당 메소드에 대한 테스트 범위인지 파악하기 더 수월할 것 같아요.

> 혹시 현업에선 테스트코드가 매우 길어지는 경우 어떻게 가독성을 높일까요?

일단 저희 팀은 QA분이 따로 계시기도 하고, 도메인당 시니어 1~2분씩만 계셔서 테스트 코드 작성은 생략하는 관계로 확인이 힘들 것 같네요 ㅠ_ㅠ
___
테스트 코드 컨벤션도 조금씩 정리해서 Readme에 주기적으로 업데이트 하면 좋을 것 같네요. 이번 주 주말 중에 한 번 관련해서 PR 올려보도록 하겠습니다.

seonpilKim commented on src/test/java/cloneproject/Instagram/domain/member/repository/querydsl/MemberRepositoryQuerydslTest.java:
```diff
@@ -0,0 +1,359 @@
+		assertThat(userProfileResponse.isFollower()).isFalse();
+		assertThat(userProfileResponse.isBlocking()).isFalse();
+		assertThat(userProfileResponse.isBlocked()).isFalse();
+	}
```
모든 필드가 매핑되는 것을 확인하는데, 빠진 필드들이 몇 개 있네요😅
name, website, image, introduce도 같이 확인하는게 확실한 품질 보증이 될 것 같습니다.

seonpilKim reviewed (approved):
리뷰가 너무 길어지는 것 같아서 몇몇 부분만 더 말씀드리고, 여기서 승인하도록 하겠습니다.
> p.s) 다음부터는 테스트 파일 하나씩 PR 올려주시는게 좋을 것 같습니다.

#### 1. 메소드 명(상태, 기대결과)이 직관적으로 느껴지지 않습니다😢 (이해하는데 어느 정도 시간이 필요합니다.)
예를 들면, `~~_MemberIsBlocking_~~` -> 직역해보면 "회원이 차단하고 있는 중이다" 인데, 코드 내부를 보면 이해가 가지만, 메소드 네이밍만 봐서는 정확히 무슨 의미인지 파악하기 조금 힘든 것 같아요. 
차라리 `~~_MemberBlockedOne_~~` 이나, 예시를 들어서 `~~_TomBlockedJerry_~~` 이런 식으로 네이밍하면 더 직관적일 것 같습니다.

#### 2. `~~_ValidArguments_AllFieldsMappedSuccess` 와는 별개로, 메소드마다 발생할 수 있는 모든 시나리오를 테스트하는게 명확하게 품질 보증이 가능할 것 같습니다.
예를 들면, `findUserProfile_ValidArguments_AllFieldsMappedSuccess`에서 `assertThat(userProfileResponse.isBlocking()).isFalse();` 를 테스트하였으니 `findUserProfile_MemberIsBlocking_FindUserProfileWithCountHidden`만 테스트 하면 되겠지, 라는 생각은 코드를 작성한 사람의 입장에서만 이해가 되는 상황이고, 제 3자 입장에서는 이런 의문이 들 수 있습니다.
> "나를 차단한 회원의 프로필을 조회하면, isBlocked가 true고, post|follower|following의 count를 볼 수 없음이 보장되겠군, 근데 나를 차단하지 않은 회원의 프로필을 조회하면 isBlocked가 false고, post|follower|following의 실제 count를 조회할 수 있음이 보장되어 있는 건가?"

 `~~_ValidArguments_AllFieldsMappedSuccess`는 dto 조회가 성공적으로 매핑되어 잘 동작하는지를 테스트하는 메소드이고,
위에서 테스트한 내용과는 별개로 각 시나리오별로 나누어서 테스트하는 것이 확실한 품질 보증을 할 수 있다고 생각합니다.

사실 저희가 querydsl을 이용해서 쿼리가 간단해 보여도, 실제 날리는 쿼리를 까보면 서브쿼리도 엄청 많고 되게 복잡한 쿼리가 되는데, 이를 테스트하는 건 엄청나게 많은 시나리오를 고려해야 할 수 밖에 없는 것 같아요.

저희가 배포된 사이트에서 이것 저것 해보면서 여러 시나리오 테스트를 진행하는 것처럼, 테스트 코드 작성 시에도 각 시나리오별로 나누어서 모든 테스트를 작성하는게 단위 테스트 작성의 의의에 맞는 것이 아닐까 라는 생각이 듭니다.

#### 번외) 리팩토링 제안
**UserProfileResponse.java**
```java
private static final int MAX_PROFILE_FOLLOWING_MEMBER_FOLLOW_COUNT = 3;
...
	public void setFollowingMemberFollow(int followCountOfMyFollowingMembers) {
		this.followingMemberFollowCount = Math.min(MAX_PROFILE_FOLLOWING_MEMBER_FOLLOW_COUNT, followCountOfMyFollowingMembers);
	}
```
**MemberService.java**
```java
	public MiniProfileResponse getMiniProfile(String username) {
		...
		final int followCountOfMyFollowingMembers = followRepository.findFollowingMemberFollowList(memberId, username).size();

		result.setFollowingMemberFollow(followCountOfMyFollowingMembers);
		...
	}
```

vectorch9 committed:
Test: 테스트내에 누락된 필드 검증 추가

- 일부 변수명 변경(isPresent로)
- 메서드명 수정

vectorch9 committed:
Test: MemberRepositoryQuerydslTest내 멤버 명에 고유명사를 사용해 구분하도록 수정
